web/src/compose_reply.js

```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
     });
 }
 
+function combine_bounding_rects(combined_rect, new_rect) {
+    if (combined_rect === null) {
+        return new_rect;
+    }
+    return {
+        top: Math.min(combined_rect.top, new_rect.top),
+        right: Math.max(combined_rect.right, new_rect.right),
+        bottom: Math.max(combined_rect.bottom, new_rect.bottom),
+        left: Math.min(combined_rect.left, new_rect.left),
+    };
+}
+
+function extract_range_html(range, preserve_ancestors = false) {
+    const temp_div = document.createElement("div");
+    if (!preserve_ancestors) {
+        temp_div.append(range.cloneContents());
+        return temp_div.innerHTML;
+    }
+    // Preserve the structure of the html upto 2 levels of ancestors.
```

Comment:
Can you extend this comment to say you're doing that?

More generally I think all of these functions would benefit from more extensive comments explaining the algorithm, and what the parameters mean.